polygeist/lib/Conversion/PolygeistToLLVM/PolygeistToLLVM.cpp

```diff
@@ -54,6 +54,362 @@ namespace mlir {
 #undef GEN_PASS_DEF_CONVERTPOLYGEISTTOLLVM
 } // namespace mlir
 
+// FIXME: All the following patterns with an "Old" suffix to their name should
+// be removed once we drop typed pointer support.
+struct BaseSubIndexOpLoweringOld : public ConvertOpToLLVMPattern<SubIndexOp> {
```

Comment:
I am thinking that we might be using non-opaque ptr for a while. Duplicating the code force us to dual maintain the code. My preference would be to keep one version of the code and use the new option introduced in this PR to change code snippets locally where we need to. WDYT?

Reply:
My motivation to use this approach (for this PR) was that we have some changes that are not as localized as a simple `if`, for example [this change here](https://github.com/intel/llvm/pull/8767/commits/235f911407c2189034635cf579f19a02e3ac703e#diff-c49fbe1ad98fd477ae47be29989f0179b5efe97eddc9e9c3d4ca6ded24cc26d1L56-L74). 
I also think that with this approach, it will be easier and less error-prone to remove typed pointer support in the future, as it will be sufficient to remove the old patterns and the logic to add them to the pass. 

From the git history, it also seemed to me that this file is not as frequently subject to changes as other files in the project. Before my changes, it hadn't been modified in almost a month (apart from a fix for the merge from `sycl`) and mainly for refactoring purposed.

For the next part of the changes (SYCL to LLVM lowering), I've chosen an approach close to what you suggest, because the changes are more localized than in this PR. 

I'm also a bit more optimistic when it comes to switching the default to opaque pointers, based on comments like [this](https://github.com/intel/llvm/pull/8286):
> All tools are switched to opaque pointers already except those we use in compilation for offload targets.

Reply:
If we are able to switch to using opaque pointers relatively quickly we can live with some code duplication for a few weeks. 
We just have to be careful and remember to dual maintain the duplicated code if we fix problems in that part of the code. 
I am OK with the PR overall.

Reply:
Agreed, we should remember that for PR reviews affecting these patterns.